File: csci40/lec29/Counter.java
public class Counter {
  // there is no such thing as function/method declarations
  // you also can't say public: or private:
  //   (so you have to put public/private next to everything)

  // we'll need one field called count
  // count is private
  private int count;

  // constructor
  public Counter() {
    // set count to 0
    count = 0;
  }

  // we can't mark a method as const very easily in Java
  public int getCount() {
    return count;
  }

  public void inc() {
    count++;
  }

  // Java doesn't have default parameters, so you have to make two
  // separate reset methods
  public void reset() {
    count = 0;
  }

  public void reset(int newCount) {
    count = newCount;
  }

  // c.add(c2)
  public Counter add(Counter other) {
    int sumCount = count + other.count;
    Counter res = new Counter();
    res.reset(sumCount);
    return res;
  }

  public static void main(String[] args) {
    Counter c = new Counter(); 
    System.out.println(c.getCount()); // 0
    c.inc();
    c.inc();
    c.inc();
    System.out.println(c.getCount()); // 3
    c.reset();
    c.inc();
    c.inc();
    System.out.println(c.getCount()); // 2
    c.reset(42);
    c.inc();
    c.inc();
    c.inc();
    System.out.println(c.getCount()); // 45

    Counter c2 = new Counter();
    c2.inc();
    c2.inc();
    c2.inc();

    Counter c3 = c.add(c2);
    System.out.println(c3.getCount()); // 48
  }
}

File: csci40/lec29/counter.py
class Counter:
    # constructors in python are called __init__
    # every method takes an explicit "self" parameter first
    # (but it's automatically passed when you call a method)
    def __init__(self, count=0):
        # let's make a field called "count"
        # all fields get made in the constructor
        self.count = count # <-- the member var gets made right here! 

    def inc(self):
        self.count += 1

    def getCount(self):
        return self.count

c = Counter()
c.inc()  # gets translated into inc(c)
c.inc()
c.inc()
print(c.getCount()) # 3

c2 = Counter(42)
c2.inc()
c2.inc()
print(c2.getCount()) # 44
